sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py

```diff
@@ -0,0 +1,158 @@
+            if start < 0:
+                start += self.length
+            if stop < 0:
+                stop += self.length
+
+            # Clamp normalized indices to valid range
+            start = max(0, min(self.length, start))
+            stop = max(0, min(self.length, stop))
+
+            # Bounds checking after normalization
+            if start > stop:
+                return ""  # Return empty string for an empty slice
+
+            # Construct region string
+            region_str = f"{self.seqid}:{start + 1}-{stop}"  # +1 for 1-based indexing
+            return self.reader.read_sequence_mmap(region_str)
+
+        elif isinstance(key, int):
+            # Normalize single integer for negative indexing
+            if key < 0:
+                key += self.length
+
+            # Bounds checking
+            if key < 0 or key >= self.length:
+                raise IndexError(f"Position {key} is out of bounds for '{self.seqid}' with length {self.length}.")
+
+            # Query single nucleotide by creating a 1-length region
+            region_str = f"{self.seqid}:{key + 1}-{key + 1}"  # +1 for 1-based indexing
+            return self.reader.read_sequence_mmap(region_str)
+
+        else:
+            raise TypeError("Index must be an integer or a slice.")
+
+
+class NvFaidx:
+    """NvFaidx is a rest + pyo3 replacement for PyFaidx that provides a dictionary-like interface to reference genomes.
+
+    NvFaidx is built using Noodles as a backend for Fai objects, and memory maps for backing the underlying fasta.
+    Using a backend of Memmaps provide the following benefits:
+        - The kernel implements this mechanism by using page faults
+        - Each read in a mmap'd file results in a page fault: there's nothing in memory to read!
+        - The kernel handles this page fault by going to the disk, reading the file in the specified offset + index,
+            then returning to the user process with what it just read, preventing penalties from context switching.
+
+    *Context*: PyFaidx or _any_ buffered read based index is not process safe, and therefore does not play nice with pytorch dataloaders.
+    Due to the order of operations, the underlying file handle is shared between processes, when `seek()` is called to perform random lookups,
+    this can cause unexpected behavior in the forked processes.
+    Ref: https://github.com/mdshw5/pyfaidx/issues/211
+
+    For a good solution we need three things:
+        1) Safe index creation, in multi-process or multi-node scenarios, this should be restricted to a single node
+            where all workers block until it is complete (not implemented above)
+        2) Index object instantion must be fast.
+        3) Read-only use of the index object must be both thread safe and process safe with python.
+    """
+
+    def __init__(self, fasta_path: str | Path, faidx_path: Optional[str | Path] = None, ignore_existing_fai=True):
```

Comment:
Nit: Should this be `str | path`, or some variant of `Union[str, Path, None]` for both fasta_path and faidx_path?

Reply:
nope! if `faidx_path` is None and `ignore_existing_fai` is True, its perfectly valid, the index is built online and stored in memory in this case. (this is in the docstring too I believe). 

But fasta_path must be set.